File: Day-4/labs/python-kafka-consumer-lab/slow_consumer.py
"""Lab 04 - slow consumer to build lag."""
import sys
import time
from kafka import KafkaConsumer
from config import BOOTSTRAP_SERVERS, LAG_TOPIC


def main():
    group_id = sys.argv[1] if len(sys.argv) > 1 else "lag-group"
    sleep_ms = int(sys.argv[2]) if len(sys.argv) > 2 else 500
    name = sys.argv[3] if len(sys.argv) > 3 else "SlowConsumer"
    consumer = KafkaConsumer(
        LAG_TOPIC,
        bootstrap_servers=BOOTSTRAP_SERVERS,
        group_id=group_id,
        client_id=name,
        auto_offset_reset="earliest",
        enable_auto_commit=False,
        value_deserializer=lambda v: v.decode("utf-8"),
    )
    print(f"[{name}] group={group_id} sleep={sleep_ms}ms")
    try:
        for msg in consumer:
            time.sleep(sleep_ms / 1000.0)
            print(f"[{name}] P={msg.partition} offset={msg.offset} value={msg.value}")
            consumer.commit()
    finally:
        consumer.close()


if __name__ == "__main__":
    main()

File: Day-4/labs/java-kafka-consumer-lab/src/main/java/com/kafka/consumer/lab/SlowConsumer.java
package com.kafka.consumer.lab;

import java.time.Duration;
import java.util.Collections;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;

/** Lab 04 - intentional delay per message to build consumer lag. */
public final class SlowConsumer {

    public static void main(String[] args) throws InterruptedException {
        String groupId = ConsumerConfigFactory.arg(args, 0, "lag-group");
        int sleepMs = ConsumerConfigFactory.parseInt(ConsumerConfigFactory.arg(args, 1, "500"), 500);
        String label = ConsumerConfigFactory.arg(args, 2, "SlowConsumer");

        var props = ConsumerConfigFactory.baseProps(ConsumerConfigFactory.DEFAULT_BOOTSTRAP, groupId);
        props.put(org.apache.kafka.clients.consumer.ConsumerConfig.CLIENT_ID_CONFIG, label);

        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
            consumer.subscribe(
                    Collections.singletonList(ConsumerConfigFactory.LAG_TOPIC),
                    new AssignmentLogger(label));

            System.out.printf("[%s] group=%s topic=%s sleep=%dms%n",
                    label, groupId, ConsumerConfigFactory.LAG_TOPIC, sleepMs);

            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(500));
                for (ConsumerRecord<String, String> record : records) {
                    Thread.sleep(sleepMs);
                    System.out.printf("[%s] P=%d offset=%d value=%s%n",
                            label, record.partition(), record.offset(), record.value());
                }
                if (!records.isEmpty()) {
                    consumer.commitSync();
                }
            }
        }
    }
}
